shared/settings.cpp

```diff
@@ -0,0 +1,102 @@
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <QDebug>
+#include "settings.h"
+
+char* readFile(const char* path) {
+    FILE *file = fopen(path, "r");
+    if (file) {
+        fseek(file, 0, SEEK_END);
+        long length=ftell(file);
+        fseek (file, 0, SEEK_SET);
+        char *buffer = (char*)malloc(length+1);
+        if (buffer) {
+            fread (buffer, 1, length, file);
+            buffer[length] = '\0';
+        }
+        fclose(file);
+        return buffer;
+    }
+    return nullptr;
+}
+
+bool doesStartWithPrefixes(const char* const prefixes[], int arrLength, char* toCheck) {
+    int length = strlen(toCheck);
+    for (int i = 0; i < arrLength; ++i) {
+        int prefixLength = strlen(prefixes[i]);
+        if (prefixLength < length && strncmp(prefixes[i], toCheck, prefixLength) == 0) {
+            return true;
+        }
+    }
+    return false;
+}
+
+void Settings::readDeviceType() {
+    char* device_type = readFile("/sys/devices/soc0/machine");
+    _deviceType = DeviceType::UNKNOWN;
+
+    if (device_type) {
+
+        const char* const remarkable1[] {
+            "reMarkable 1",
+            "reMarkable Prototype 1",
+        };
+        const char* const remarkable2[] {
+            "reMarkable 2"
+        };
+
+        if (doesStartWithPrefixes(remarkable1,
+              sizeof(remarkable1) / sizeof(remarkable1[0]), device_type)) {
+            _deviceType = DeviceType::RM1;
+            qDebug() << "RM1 detected...";
+        }
+        else if (doesStartWithPrefixes(remarkable2,
+              sizeof(remarkable2) / sizeof(remarkable2[0]), device_type)) {
+            _deviceType = DeviceType::RM2;
+            qDebug() << "RM2 detected...";
+
+        }
+        free(device_type);
+    }
+}
+
+DeviceType Settings::getDeviceType() {
+    if (_deviceType == DeviceType::UNDETERMINED) {
+        readDeviceType();
```

Comment:
It would make more sense to read the device type as part of the constructor.

Reply:
Doing IO in the constructor is generally not the best idea, but whatever.

Reply:
Do you have some links to places that talk about why it's not a good idea?

Reply:
It will make it untestable without the file, for example. But it's simple enough and there are not tests, so lets read the data in the constructor.